### ISSUE TITLE
GGC Weapon Refinement

### ISSUE THREAD

**Issue by LeeBamberTGC** [TheGameCreators/GameGuruRepo#1406]
THE ORIGINAL REQUEST:

7) Want to be able to specify minimum ammo available before being able to reload (feature request).
This is primarily for weapons where multiple rounds of ammo can be clearly seen when reloading.  This feature would help to prevent visual continuity issues.  Particularly when reloading while the player has only a very small amount of ammo in their inventory. The minimum reload amount should be specified via gunspec with something like "minreloadqty=x" where the default is 1 and any other specified number is the combined value of ammo stored currently in the gun plus ammo held in ammo pool.  So, if minreloadqty=15 then the number of bullets in the gun plus bullets in their pocket must equal 15 or more.  If that value is less than 15 then they simply can't initiate a reload.   Arguably, there is a negative gameplay tradeoff if the ability to reload/make use of a weapon is gate kept in this manner.  As such, it's something that could be easily disabled by the game dev. should they choose to prefer the default method.

8) Need to be able to specify that a weapon is reloaded as a result of alt mode toggling (feature request).
When a weapon changes ammo type via alt mode switching, there is a high probability that it will involve changing which ammo is loaded into the weapon. Such is the case with the LMG where one type of ammo is loaded into the top via belt and the other ammo type is loaded into the side via magazine.  This can be observed/tested by toggling between fire modes [V-key] when one the modes are in different combinations of loaded or empty states. Currently, you can toggle from either mode while empty to a non-empty mode and it works fine but if you toggle to a mode that is empty, the animation does not mate up and the visual activity is nonsensical. In this scenario, when switching back and forth between modes, it makes sense that the ammo is topped up for the mode being switched to.  Likewise, it also makes sense that alt mode switching should not be togglable if no ammo is available for the mode being switched to (this kind of is related to #7 above and represents a separate issue that I don't really have any concrete solutions for since I don't know how the code works).
Currently, we do not have animation support for the different combinations of empty/non-empty to empty state animations.  I do not think we should add support for every state possibility because it needlessly adds animation data and wastes the player's time without benefit and detracts from immersion. We currently support the animations: alto, altfrom, empty altto. and empty altfrom. This will be sufficient if a weapon can be flagged to reload it's ammo on alt switch because empty altto and empty altfrom can be animated in a way where it transitions seamlessly from an empty state to a full state as part of the alt toggle.  Essentially, there just needs to be a way to flag a weapon so that it performs the reload logic upon initiating an alt toggle. There are probably a few different ways to go about this.  One way would be to add a new gunspec setting like "reloadalt = x" where default is 0 and does not reload and a value of 1 reloads.  Another way to do it may be to expand the capabilities of the "shareloadedammo=x" setting.  Currently, shareloadedammo=1 shares loaded/in-gun ammo qty. and =0 does not.  If this was extended so that shareloadedammo=2 meant that ammo is not shared and ammo is replenished (to the target mode) on mode change. The second suggestion probably sounds more convoluted but may prove to be more robust if additional logic needs to be added in the future (GGM) for unforeseen alt/ammo handling scenarios (ie: is it preferable to have an additional field/checkbox or an additional dropdown option).

THE TWEAKS:

For 7, it should block if the total available between what's in the pool combined with what is already in the gun is less than the minreloadqty.
So if:
minreloadqty = 15
Then...
10 in pool + 5 in gun = YES
10 in pool + 0 in gun = NO
If it's already like this then we're good.
  
For 8, If the target mode was in empty state when last used, upon switching modes, it will briefly show the empty mode idle then snap to normal idle causing a jarring visual glitch.  I'm guessing that it's because the engine is remembering that the target mode was in empty state and rechecking only after the altto or altfrom animation has completed and returned to idle. As a sort of lateral matter, I will need to test how 7 and 8 play together.
